jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,335 @@
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = new LinkedList<>();
+                newLabeledPath.vertices.addAll(this.vertices);
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    private Graph<V, E> graph;
+    private Map<V, Integer> labels;
+    private int lengthBound;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = graph;
+        this.lengthBound = lengthBound;
+        this.labels = labels;
+    }
+
+    /**
+     * calculates a valid subset-disjoint negative cycle.
+     * If there is no such a cycle, it returns an empty LabeledPath instance with cost Double.MAX_VALUE.
+     *
+     * @return a valid subset-disjoint negative cycle encoded as LabeledPath
+     */
+    public LabeledPath<V> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> cycleResult = new LabeledPath<>();
+
+        Set<LabeledPath<V>> pathsLengthK = new LinkedHashSet<>();
+        Set<LabeledPath<V>> pathsLengthKplus1 = new LinkedHashSet<>();
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
+            if(graph.getEdgeWeight(e) < 0) {
+                // initialize all paths of cost < 0
+                LinkedList<V> pathVertices = new LinkedList<>();
+                Map<V, Integer> pathLabels = new LinkedHashMap<>();
+                V sourceVertex = graph.getEdgeSource(e);
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.put(sourceVertex, labels.get(sourceVertex));
+                pathLabels.put(targetVertex, labels.get(targetVertex));
+                LabeledPath<V> path = new LabeledPath<>(pathVertices, graph.getEdgeWeight(e), pathLabels);
+
+                // add path to set of paths of length 1
+                pathsLengthK.add(path);
+            }
+        }
+
+        while(k < lengthBound && cycleResult.getCost() >= 0) {
+            while(!pathsLengthK.isEmpty()) {
+
+                // go through all valid paths of length k
+                for(Iterator<LabeledPath<V>> it = pathsLengthK.iterator(); it.hasNext();) {
```

Comment:
Why do you need to remove the paths iteratively?  You're just going to discard the entire pathsLengthK at the end anyway.  Skip that and then use a compact for loop `for (LabeledPath<V> path : pathLengthK)`.